lib/edgehog_device_forwarder/connections/token_data.ex

```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Connections.TokenData do
+  defstruct realm: nil, device_id: nil, socket: nil
+
+  @type t :: %__MODULE__{}
+  @type device_id :: String.t()
+  @type realm :: String.t()
+  @type socket :: pid()
+
+  @spec new(socket) :: t
+  def new(socket), do: %__MODULE__{socket: socket}
+
+  @spec put_device_info(nil | t, %{:device_id => device_id, :realm => realm}) ::
+          {:ok, t} | {:error, :token_not_found}
+  def put_device_info(nil, _device_info), do: {:error, :token_not_found}
+
+  def put_device_info(data, device_info) do
+    updated_data = %__MODULE__{
+      realm: device_info.realm,
+      device_id: device_info.device_id,
+      socket: data.socket
+    }
+
+    {:ok, updated_data}
```

Comment:
Instead of constructing a new map, since you want to keep `socket`, you can update the existing one with the update syntax:
```elixir
{:ok, %{data | realm: device_info.realm, device_id: device_info.device_id}}
```